File: gred/src/undo_manager.lua
UndoManager = class()

function UndoManager:Create()
  self = self:New()
  self._undo = {}
  self._redo = {}
  return self
end

function UndoManager:addUndo(command)
  if command ~= nil then
    self._undo[#self._undo + 1] = command
    self._redo = {}
  end
end

function UndoManager:canUndo()
  return #self._undo > 0
end

function UndoManager:canRedo()
  return #self._redo > 0
end

function UndoManager:undo()
  if self:canUndo() then
    self._redo[#self._redo + 1] = self._undo[#self._undo]:execute()
    self._undo[#self._undo] = nil
  end
end

function UndoManager:redo()
  if self:canRedo() then
    self._undo[#self._undo + 1] = self._redo[#self._redo]:execute()
    self._redo[#self._redo] = nil
  end
end

function UndoManager:update()
  if IsKeyDown(KEY_LEFT_CONTROL) or IsKeyDown(KEY_RIGHT_CONTROL) then
    if IsKeyPressed(KEY_Z) then self:undo() end
    if IsKeyPressed(KEY_Y) then self:redo() end
  end
end

function UndoManager:reset()
  self._undo = {}
  self._redo = {}
end

File: src/undo_manager.cpp
#include "undo_manager.h"
#include "app.h"

UndoManager::UndoManager() {}

void UndoManager::add_undo(std::shared_ptr<ICommand> cmd) {
    if (cmd) {
        undo_stack.push_back(cmd);
        redo_stack.clear();
    }
}

bool UndoManager::can_undo() const { return !undo_stack.empty(); }
bool UndoManager::can_redo() const { return !redo_stack.empty(); }

void UndoManager::undo() {
    if (!can_undo()) return;
    auto result = undo_stack.back()->execute();
    undo_stack.pop_back();
    if (result) redo_stack.push_back(result);
}

void UndoManager::redo() {
    if (!can_redo()) return;
    auto result = redo_stack.back()->execute();
    redo_stack.pop_back();
    if (result) undo_stack.push_back(result);
}

void UndoManager::update() {
    bool ctrl = App::key_down[KEY_LCONTROL] || App::key_down[KEY_RCONTROL];
    if (ctrl) {
        if (App::key_hit[KEY_KEY_Z]) undo();
        if (App::key_hit[KEY_KEY_Y]) redo();
    }
}

void UndoManager::reset() {
    undo_stack.clear();
    redo_stack.clear();
}
